proxy/round_tripper/proxy_round_tripper.go

```diff
@@ -233,13 +236,39 @@ func (rt *roundTripper) backendRoundTrip(
 
 	rt.combinedReporter.CaptureRoutingRequest(endpoint)
 	tr := GetRoundTripper(endpoint, rt.roundTripperFactory)
-	res, err := tr.RoundTrip(request)
+	res, err := rt.timedRoundTrip(tr, request)
 
 	// decrement connection stats
 	iter.PostRequest(endpoint)
 	return res, err
 }
 
+func (rt *roundTripper) timedRoundTrip(tr http.RoundTripper, request *http.Request) (*http.Response, error) {
+
+	if rt.endpointTimeout <= 0 {
+		return tr.RoundTrip(request)
+	}
+	reqCtx, cancel := context.WithTimeout(request.Context(), rt.endpointTimeout)
```

Comment:
This bit is interesting due to our use of `reverse-proxy`. The go stdlib has already put a cancellation context onto the request that has been passed into this function. You'll see that cancellation get wired up here:

https://github.com/golang/go/blob/master/src/net/http/httputil/reverseproxy.go#L144-L165

That would potentially mean that `WithTimeout` could potentially instead be a `time.After` with the specified `rt.endpointTimeout` which would close the context `Done` chan if the timeout was reached. I think that may potentially work but the same but have not tried it.

Reply:
@zachgersh The problem is that the gorouter relays most of the work about handling the request to the reverse proxy. The reason we have to set the `WithTimeout` is because only the reverse proxy knows when the request is finished. So even if we do a `time.After`, there is no capability for us to be able to know whether the request is finished to be able to cancel the request or not.